### PR TITLE
Blog Post: Benchmarking Kotlin (Http4k, Ktor) and Rust (Actix) Microservices

### PR DESCRIPTION
**Pull request by strohel**
I will be happy for review of facts, style, English, anything & everything.

I fear the post is a bit too long, so any suggestions for shortening, simplification and omissions welcome.

The images don't render inline in GitHub rich diff, but they can be seen on: https://storage.googleapis.com/strohel-pub/bench-3-impls/bench-results.html (in order)

CC @goodhoko @knyttl @lengerad.

### CODE REVIEW

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 1]
```diff
@@ -0,0 +1,445 @@
+---
+title: Benchmarking Kotlin (Http4k, Ktor) and Rust (Actix) Microservices
```
Comment (strohel) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Allternative:
`Benchmarking Http4k, Ktor (Kotlin) and Actix (Rust) Microservices`
Better :+1:  or worse :-1: ?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+#image: /images/2020-09-06-bench-rust-kotlin-microservices/TODO
+#urlo: TODO
+#reddit: TODO
+#hn: TODO
+---
+
+We were looking for a next microservice framework to power our future[^framework] lightweight web
+services back in spring 2020 at [GoOut][goout].
+Apart from detailed (and sometimes philosophical) theoretical comparisons that I much enjoyed,
+we decided to implement proof-of-concept microservices in most viable ones,
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: ...microservices using the most viable ones... (maybe it sounds better)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+which is touted for high performance.
+I ended up stress-testing Kotlin's [**Http4k**][http4k], [**Ktor**][ktor][^vertx],
+and Rust's [**Actix**][actix], read on to see how they fared.
+
+[goout]: https://goout.net/
+[http4k]: https://www.http4k.org/
+[ktor]: https://ktor.io/
+[actix]: https://actix.rs/
+
+[^framework]: The existing framework being Spring servlets written in Kotlin running in Tomcat.
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: was/is/were/are  instead of being Spring servlets

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+### Recipe
+
+The benchmark consists of multiple identical "runs" for each Dockerised implementation,
+see [bench-3-impls.py](https://github.com/strohel/locations-rs/blob/master/bench-3-impls.py).
+These are named after the implementation and ordinal number of the run,
+thus `rs-actix-1` is the first run of the Rust (Actix) implementation,
+and `kt-ktor-3` is the third run of Kotlin (Ktor) implementation.
+
+Each run simulates first minutes of the microservice instance, see [test-image.py](https://github.com/strohel/locations-rs/blob/master/test-image.py). Steps of one run:
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: first couple of minutes?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+1. Docker container is started, measuring time until the service starts responding,
+2. a series of HTTP checks is made ensuring correct responses and error handling behaviour,
+3. HTTP load is generated using [wrk](https://github.com/wg/wrk)
+   in a series of 16 10-second steps with increasing concurrent connection count;
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: 10-second long steps

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+   the metrics are collected from `wrk` invocation and Docker API after each step:
+   - first 5 steps use just one concurrent connection, acting as a warm-up,
+   - each of the next 11 steps doubles concurrent connection count, i.e. continuing with 2, 4,
+     8, ..., 1024, 2048.[^conn-count]
+
+[^conn-count]: The top-end concurrent connection counts of 1024 and 2048 are beyond practical utility.
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: (subjective) `beyond practical utility` does not sound right to me, I would use a different figure of speech

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+
+The startup time is measured from the moment Docker finishes setting up the container
+to the moment when we receive a valid HTTP response to `GET /`.
+The services are required not to start if configured Elasticsearch server is unreachable,
+i.e. required to do an HTTP ping round-trip to it.
+
+Both JVM-based Kotlin services start well under 3 seconds,
+which is impressive given the JVM has to load itself, service 40+ MiB fat JAR and initialise it.
+
+Actix's Rusty startup times are barely visible in the graph as they take 1--2 ms,
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: wow

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+
+Here we measure high-water mark memory usage of the container from its start till the end of each
+benchmark step, as reported by Docker (i.e. not just momentary memory usage).
+
+The comparison between JVM-based and Rust frameworks is not really representative here
+as allocated Java heap depends on configuration in addition to actual usage.
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/max_mem_usage_per_requests.svg" />
+
+The same metric as above, but divided by successful requests per second;
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: by the number of

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+which gives unconventional unit of *megabyte-seconds per request*.
+
+Higher initial values are driven by the constant portion of memory that each framework requires.
+Afterwards, Http4k nicely settles at flat value of ~0.05 MiB⋅s in spite of increasing latencies,
+which may indicate well-tuned request backlog handling.
+
+Async frameworks Ktor and Actix do exhibit some increase of time-integrated per-request memory usage
+with growing number of connections.
+This may indicate they are trying to be a bit too much of concurrent there
```
Comment (spetex) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: bit too much of concurrent? I dont get this.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 1 above)
Comment (spetex) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: better :+1:

**Review by spetex — COMMENTED** [review]
Very good, I like it. I did not even knew you were doing such things at the time. I knew you were doing some comparisons but I thought it ended up with discussions.

For the blog post: It seems in places to be a bit too hard to read, its very information dense and it feels like reading a science paper rather a blog post. You might want to consider the tone. If that is desired its ok, but If you want more people to read it, I would probably add more sauce and less data. I know you wanted to present all of it but the scale between readability and information presented is hard to balance.

Other than that its very interesting blog post.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+Line labels on the graph correspond to [tags in the locations-kt-http4k repository](https://gitlab.com/gooutopensource/locations-kt-http4k/-/tags).
+
+One nice side-effect of this benchmarking was that
+I have [discovered a cause](https://github.com/http4k/http4k/issues/141#issuecomment-679330954)
+of the [**Netty** backend performance issue](https://github.com/http4k/http4k/issues/141),
+and [@**daviddenton** was quick to fix it](https://github.com/http4k/http4k/pull/480).
+The `netty-updfix-*` graphs, which exhibit some nice latency characteristics, already contain the fix.
+It was later released in http4k v3.259.0 (be sure to upgrade if you use this backend).
+
+**Apache HttpCore** version 4.4, labelled as `apache4-*`, was most CPU-efficient backend
```
Comment (ttichacek) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: `apache4-*` apod. Ja jsem asi pitomej, ale nejak jsem nedokazal rozklíčovat ten placeholder `*`.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 11]
```diff
@@ -0,0 +1,445 @@
...
+The **Jetty** backend (`jetty-*`) slightly but consistently lags behind other production engines
+in most metrics.
+For completeness I have also included the **SunHttp** backend that is not intended for production use,
+which the benchmarks clearly confirm.
+
+### The Story of `locations-kt-ktor`
+
+Ktor variant was also written by [@**goodhoko**][goodhoko] at [GoOut][goout],
+I later tweaked the `/city/v1/get` endpoint for functional parity with other implementations.
+It remains least complete implementation with just this single endpoint.
```
Comment (ttichacek) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]:
```suggestion
It remains the least complete implementation with just this single endpoint.
```
?

**Comment by strohel** [comment]
@spetex wrote:
> For the blog post: It seems in places to be a bit too hard to read, its very information dense and it feels like reading a science paper rather a blog post. You might want to consider the tone. If that is desired its ok, but If you want more people to read it, I would probably add more sauce and less data. I know you wanted to present all of it but the scale between readability and information presented is hard to balance.

That's what I feared. Good point, I'll see what I can improve. I definitely want this to be for wider (technical) audience. http://www.hemingwayapp.com/ suggest splitting a lot of sentences, so maybe that could help a bit.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 3 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Or maybe I can remove this footnote altogether. Do you, reviewers, think it adds something for the reader?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 6 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: `... are too high to be practical`?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 10 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Hmmm, to má pojednávat o těch labelech na https://storage.googleapis.com/strohel-pub/bench-http4k-server-engines-pub/bench-results.html - jakože `apche4-1`, `apache4-2`, ...

Ono je trochu slabý místo toho blog postu že tady rozebírám něco, co je vidět jen po kliknutí na to `See Kttp4k server engine results on a dedicated page`.

Možná tam šoupnout preview obrázek, aby to bylo jasnější?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 3 above)
Comment (spetex) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: It adds background, I would keep it. Maybe you can just avoid foot note and say spring right in the sentence?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 6 above)
Comment (spetex) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: :+1:

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 9 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: I wanted to convey my theory that async frameworks do try to start processing all requests as they arrive (which holds memory for a longer time), difference to sync framework. I'll try to reword this in the next iteration.

Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Rewritten in new revision, hopefully better.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+2. a suite of HTTP checks is made ensuring correct responses and error handling behaviour,
+3. HTTP load is generated using [wrk](https://github.com/wg/wrk)
+   in a series of 10-second long steps with increasing concurrent connection count;
+   the metrics are collected from `wrk` and Docker API after each step:
+   - first 5 steps use just one concurrent connection, acting as a warm-up,
+   - each of the next 11 steps doubles concurrent connection count, i.e. continuing with 2, 4,
+     8, ..., 1024, 2048.[^conn-count]
+
+[^conn-count]: The top-end concurrent connection counts of 1024 and 2048 are too high to be practical.
+    They are included to test framework behaviour under uncontrolled traffic spike.
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: If you wanted the graphs be super-nice you could mark the two right-most data-columns (1024 and 2048 conns) with eg. a different background (like greyed out or similar). It would be nice to convey this remark solely with the images as many readers (IMO) will go straight to them and won't read this footnote.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+
+Finally, all runs are plotted using awesome [Pygal](http://www.pygal.org/) library,
+see [render-tests.py](https://github.com/strohel/locations-rs/blob/master/render-tests.py).
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Rather than distracting the reader by imperatively instructing him to go see something, you could be more subtle on this.

```suggestion
Finally, all runs are plotted using awesome [Pygal](http://www.pygal.org/) library
in [render-tests.py](https://github.com/strohel/locations-rs/blob/master/render-tests.py).
```

In general the blogpost is full of links and footnotes that 'force' the reader to jump away (and back). It is IMO more comfortable to read linearly without the need to branch very often.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 14]
```diff
@@ -0,0 +1,448 @@
...
+Different runs of a single implementation use the same colour,
+allowing for visual inspection of the variance between them.
+
+### Runtime Environment
+
+Docker is told to constrain resources available to the microservice containers,
+as in a typical Kubernetes cluster.
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: source? .)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 15]
```diff
@@ -0,0 +1,448 @@
...
+
+3. `locations-rs`: [code on GitHub](https://github.com/strohel/locations-rs/tree/rs-actix).
+   - lang: **Rust** v1.45.2
+   - framework: [**Actix** web][actix] v2.0
+   - programming model: **async** (Rust *async/await* with [Tokio][tokio] runtime v0.2.22 through `actix-rt`)
+
+[tokio]: https://tokio.rs/
+
+[^http4k-perf]: A newer http4k version 3.258.0 with Undertow 2.1.3 very slightly regressed in
+    performance, see http4k server engine benchmarks.
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Could you link to the mentioned `http4k server engine benchmarks`?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 16]
```diff
@@ -0,0 +1,448 @@
...
+
+### The Story of `locations-kt-http4k`
+
+Http4k Locations service was written by [@**goodhoko**][goodhoko] at [GoOut][goout].
+It became the production implementation at GoOut
+and is thus most complete along with extras like Swagger serving.
+
+[goodhoko]: https://github.com/goodhoko
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: <3! slo by takto pls?
```suggestion
[goodhoko]: https://buhvi.co
```

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 17]
```diff
@@ -0,0 +1,448 @@
...
+
+The same metric as above, but divided by the number of successful requests per second;
+which gives an unconventional unit of *megabyte-seconds per request*.
+
+It is normal that the values for low connections counts are higher
+because the base memory footprint of each framework is spread out across fewer requests.
+
+You may notice a slight difference in shape between Http4k and the 2 async frameworks:
+Http4k gets down to ~0.05 MiB⋅s and stays there, while the others go up again as latencies increase.
+If you have an explanation for this, please share it in a discussion.
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: It's clear that the cause is the same for this and for the already [mentioned mystery](https://github.com/strohel/strohel.github.io/pull/2/files?short_path=7df25e1#diff-c1100ea69a8c13d3d16cf9875c68c6a2R328).

Might be worth noting in the text. .)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 18]
```diff
@@ -0,0 +1,448 @@
...
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu.svg" />
+
+Consumed CPU time as reported by Docker API.
+
+JVM is seen working hard to JIT-optimise Kotlin-based codebases during the warm-up steps.
+The just-in-time optimisation, combined with the execution of unoptimised code, takes roughly 23 CPU-seconds extra,
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: I guess you calculated the `23 s` figure by comparing the first four 1-conn runs to the fifth optimised one, right? (Just checking I understand this well .) )

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 19]
```diff
@@ -0,0 +1,448 @@
...
+
+Still, this is something to keep in mind about JVM-based microservices
+when considering, for example, high-frequency continuous deployment or aggressive instance count autoscaling.
+The JIT may compete for resources with the not-yet-optimised target workload.
+Such characteristic should also apply to other JIT-based engines like Node.js' V8.
+
+Further right in the graph, Kotlin implementations quickly saturate the allocated CPU portion,
+Rust following ~2 steps later.
+
+I'm curious that Http4k manages to consume a bit more CPU than the 15 CPU-seconds assigned by Docker.
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: maybe...?

```suggestion
I'm curious how Http4k manages to consume a bit more CPU than the 15 CPU-seconds assigned by Docker.
```

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 20]
```diff
@@ -0,0 +1,448 @@
...
+
+<embed type="image/svg+xml" src="/images/2020-09-06-bench-rust-kotlin-microservices/cpu_per_request.svg" />
+
+Consumed CPU time divided by successful requests per second.
+
+I'm not sure how to explain the improving efficiency of Rust (Actix) with the increasing number of concurrent connections.
```
Comment (goodhoko) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Followup blogpost? 😉

**Comment by goodhoko** [comment]
Overall very good. (b^_^)b

IMO the best parts are the ones where you make observations of the results, their implications and possible explanations. 

Pursuing [the desire of wider audience accesibility](https://github.com/strohel/strohel.github.io/pull/2#issuecomment-688191854) I'd suggest to make the text more linear. (As I already wrote above) Try to replace footnotes with sentences directly in the text. Maybe simplify [nested lists](https://github.com/strohel/strohel.github.io/pull/2/files#diff-c1100ea69a8c13d3d16cf9875c68c6a2R328) to flat lists. etc. .)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 14 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-resource-requests-and-limits
> While your Kubernetes cluster might work fine without setting resource requests and limits, you will start running into stability issues as your teams and projects grow. Adding requests and limits to your Pods and Namespaces only takes a little extra effort, and can save you from running into many headaches down the line!

People always follow best practices, right? ;)

Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: The article is kinda funny. A person with a position called "Developer Advocate" says:
> This could be caused by a team spinning up more replicas than they need to artificially decrease latency (hey, it’s easier to spin up more copies than make your code more efficient!), ...

and
> Regardless of whether the issue is caused by a bad developer, bad code, or bad luck, ...

;-)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 18 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Yes. Or rather by subtracting 5*(CPU time of last 1-connection step) from the actual total CPU-secs used by the container at the end of last 1-connection step. But it gives a similar number to counting the "extra height" in the graph above.

I ended up just adding the word *roughly* instead of the recipe above for the sake of conciseness. Does it sound reasonable?

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 17 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Actually, I think this is not caused by
> The behaviour of Http4k in extreme connection counts is *opposite*, reaching its peak *successful* req/s of ~3500, despite the presence of non-zero error ratio.

...because it happens more early and in "normal" connection counts zone 32 to 256. Actix goes from ~0.0011 MiB⋅s to  ~0.0037, while Http4k stays still constant (albeit much higher).

But I do acknowledge that Http4k req/s, CPU efficiency and CPU usage are tightly related. I'll reword all the "mystery" sentences including this one, let's see.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 19 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Actually, I have an idea *how* this happens: Http4k uses a lot of threads and maybe the CFS scheduler's Bandwidth Control may create a small error for each thread, which accumulates. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/scheduler/sched-bwc.rst#n54 mentions task run-time being synchronized only once per 5ms, but that synchronization between CPU cores, not between threads.

OR, http4k manages to create more *system* (kernel) load than others (again high thread count?), and perhaps Docker limits just userspace load, but includes system load portion in its stats.

Mention that or not? Inline or footnote? I have a hard time deciding whether to include potentially interesting *and* false ideas.

I'll rewrite this a bit and let's see.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 20 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Maybe, but I would have to do quite a lot of research for that. Let's see if the posts atracts some comments from more knowledgeable.

**Comment by strohel** [comment]
@goodhoko wrote:
> Pursuing the desire of wider audience accesibility I'd suggest to make the text more linear. (As I already wrote above) Try to replace footnotes with sentences directly in the text.

Agreed. It seems that I tend to abuse footnotes to balance 2 contradicting desires: not being too wordy, but at the same time include all potentially useful information. I wish footnotes would be nicely inline like in https://fivethirtyeight.com/features/a-big-chunk-of-white-americans-with-degrees-and-people-of-color-are-behind-trump/ ("majority of voters in all three.")

> Maybe simplify nested lists to flat lists. etc. .)

:+1:

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md [hunk 21]
```diff
@@ -0,0 +1,448 @@
...
+I desired to compare each framework in its best shape,
+so I ended up doing 3 qualification benchmarks,
+which may hopefully serve people also on their own.
+All implementations, tools and data are open-sourced,
+so the benchmarks should be in theory reproducible by anybody;
+also it should be easy to add new implementations to the mix.
+
+The post was only about runtime characteristics,
+which is just one of many aspects of each technology.
+Comparing developer experience would be material for a whole new write-up!
```
Comment (strohel) [inline _posts/2020-09-06-bench-rust-kotlin-microservices.md]: @goodhoko want to participate on that? :)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 14 above)
Comment (goodhoko) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: aaah, damn I understood the sentense completely differently; like
> Docker constrains resources in the same way as they're constrained in typical k8s cluster.

and not

> We tell docker to constrain resources as is typically done in a k8s cluster.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 18 above)
Comment (goodhoko) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: It does. Thoughtful reader will figure out the formula themselves. .)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 14 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Aha, then it calls for rewording

**Comment by goodhoko** [comment]
@strohel 
> I wish footnotes would be nicely inline like in https://fivethirtyeight.com/features/a-big-chunk-of-white-americans-with-degrees-and-people-of-color-are-behind-trump

Isn't that a question of merely choosing the right blog theme? ,)

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 21 above)
Comment (goodhoko) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Had I been on dev.to I could write a post comparing 3 different frameworks based on my deep experience of implementing a single endpoint in them. However, that's something I'd probably wouldn't want to present myself with. .))


Otoh I could invest my time and finish the ktor and vertx services to be feature complete and **then** write the post. But you know...
https://www.youtube.com/watch?v=HL1UzIK-flA&ab_channel=RihannaVEVO

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 20 above)
Comment (goodhoko) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: 🤞

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 21 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: Ack. Let's see if there is some feedback requesting such a post, the hint is pretty obvious.

File: _posts/2020-09-06-bench-rust-kotlin-microservices.md (hunk 12 above)
Comment (strohel) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: So I tried, but it seems I would have to [use a custom CSS for the SVGs](http://www.pygal.org/en/stable/documentation/custom_styles.html#using-a-custom-css) and that seems a bit complicated. :-|

Comment (goodhoko) [reply _posts/2020-09-06-bench-rust-kotlin-microservices.md]: 👍 for trying. I admit that's mostly unnecessary. .)

**Comment by lengerad** [comment]
Very good! 👏 

Probably just small nitpicks, 

- in the section with further description of `ktor` and especially in the [end](https://github.com/strohel/strohel.github.io/pull/2/files#diff-c1100ea69a8c13d3d16cf9875c68c6a2R245) of `rust` section I caught myself thinking that it's getting bit detailed (which is good, but as well maybe too wide for the "introduction part") and I'm already curious and excited for the graphs and the comparison by itself :)).
- also agree with lot of footnotes and references, I prefer to no open so many links or jump bottom and back too often as it distracts me (but I was also reading the downloaded PDF version, so it may already change a bit), so it would be nice to replace it with single descriptive sentence and something like "for more info, see end of the article"

great job 🙇‍♀️

**Comment by strohel** [comment]
Aaaand it's published: https://matej.laitl.cz/bench-rust-kotlin-microservices/

Thanks a bunch again for everybody's comments, the published version is so much different than the initial one.
